TypeScripts changes causes build issues

<!-- Guidelines -->
<!-- Please check if an issue does not exist already for it: https://github.com/DataDog/browser-sdk/issues -->
<!-- Please check if the issue happens with latest version -->

**Describe the bug**
Ever since version 4.40.0 (and I'm assuming the type changes that went in for it, our build system is broken. Here is the error I receive:

```
Error: node_modules/@datadog/browser-rum-core/cjs/domain/lifeCycle.d.ts(27,6): error TS2748: Cannot access ambient const enums when 'isolatedModules' is enabled.
```

**To Reproduce**
Not sure I can provide a repro case. My tsconfig for the project is
```
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "baseUrl": "src",
    "composite": true,
    "isolatedModules": true,
    "lib": ["DOM", "DOM.Iterable", "ESNext"],
    "module": "esnext",
    "moduleResolution": "node",
    "noEmit": true,
    "resolveJsonModule": true,
    "skipLibCheck": false,
    "sourceMap": false,
    "target": "ES6"
  },
```

**Expected behavior**
A breaking change like this should be done in a major version although I suspect this is a bug

--- a/packages/rum-core/src/domain/lifeCycle.ts
+++ b/packages/rum-core/src/domain/lifeCycle.ts
@@ -47,7 +47,9 @@ export const enum LifeCycleEventType {
 // "declare" statement, it will only be used during typecheck and completely ignored when building
 // JavaScript).
 //
-// See issue https://github.com/DataDog/browser-sdk/issues/2208
+// See issues:
+// * https://github.com/DataDog/browser-sdk/issues/2208
+// * https://github.com/microsoft/TypeScript/issues/54152
 declare const LifeCycleEventTypeAsConst: {
   PERFORMANCE_ENTRIES_COLLECTED: LifeCycleEventType.PERFORMANCE_ENTRIES_COLLECTED
   AUTO_ACTION_COMPLETED: LifeCycleEventType.AUTO_ACTION_COMPLETED

--- a/packages/rum-core/src/domain/lifeCycle.ts
+++ b/packages/rum-core/src/domain/lifeCycle.ts
@@ -37,22 +37,49 @@ export const enum LifeCycleEventType {
   RAW_ERROR_COLLECTED,
 }
 
+// This is a workaround for an issue occurring when the Browser SDK is included in a TypeScript
+// project configured with `isolatedModules: true`. Even if the const enum is declared in this
+// module, we cannot use it directly to define the EventMap interface keys (TS error: "Cannot access
+// ambient const enums when the '--isolatedModules' flag is provided.").
+//
+// Using a plain enum would fix the issue, but would also add 2KB to the minified bundle. By using
+// this workaround, we can keep using a const enum without impacting the bundle size (since it is a
+// "declare" statement, it will only be used during typecheck and completely ignored when building
+// JavaScript).
+//
+// See issue https://github.com/DataDog/browser-sdk/issues/2208
+declare const LifeCycleEventTypeAsConst: {
+  PERFORMANCE_ENTRIES_COLLECTED: LifeCycleEventType.PERFORMANCE_ENTRIES_COLLECTED
+  AUTO_ACTION_COMPLETED: LifeCycleEventType.AUTO_ACTION_COMPLETED
+  VIEW_CREATED: LifeCycleEventType.VIEW_CREATED
+  VIEW_UPDATED: LifeCycleEventType.VIEW_UPDATED
+  VIEW_ENDED: LifeCycleEventType.VIEW_ENDED
+  REQUEST_STARTED: LifeCycleEventType.REQUEST_STARTED
+  REQUEST_COMPLETED: LifeCycleEventType.REQUEST_COMPLETED
+  SESSION_EXPIRED: LifeCycleEventType.SESSION_EXPIRED
+  SESSION_RENEWED: LifeCycleEventType.SESSION_RENEWED
+  PAGE_EXITED: LifeCycleEventType.PAGE_EXITED
+  RAW_RUM_EVENT_COLLECTED: LifeCycleEventType.RAW_RUM_EVENT_COLLECTED
+  RUM_EVENT_COLLECTED: LifeCycleEventType.RUM_EVENT_COLLECTED
+  RAW_ERROR_COLLECTED: LifeCycleEventType.RAW_ERROR_COLLECTED
+}
+
 // Note: this interface needs to be exported even if it is not used outside of this module, else TS
 // fails to build the rum-core package with error TS4058
 export interface LifeCycleEventMap {
-  [LifeCycleEventType.PERFORMANCE_ENTRIES_COLLECTED]: RumPerformanceEntry[]
-  [LifeCycleEventType.AUTO_ACTION_COMPLETED]: AutoAction
-  [LifeCycleEventType.VIEW_CREATED]: ViewCreatedEvent
-  [LifeCycleEventType.VIEW_UPDATED]: ViewEvent
-  [LifeCycleEventType.VIEW_ENDED]: ViewEndedEvent
-  [LifeCycleEventType.REQUEST_STARTED]: RequestStartEvent
-  [LifeCycleEventType.REQUEST_COMPLETED]: RequestCompleteEvent
-  [LifeCycleEventType.SESSION_EXPIRED]: void
-  [LifeCycleEventType.SESSION_RENEWED]: void
-  [LifeCycleEventType.PAGE_EXITED]: PageExitEvent
-  [LifeCycleEventType.RAW_RUM_EVENT_COLLECTED]: RawRumEventCollectedData
-  [LifeCycleEventType.RUM_EVENT_COLLECTED]: RumEvent & Context
-  [LifeCycleEventType.RAW_ERROR_COLLECTED]: {
+  [LifeCycleEventTypeAsConst.PERFORMANCE_ENTRIES_COLLECTED]: RumPerformanceEntry[]
+  [LifeCycleEventTypeAsConst.AUTO_ACTION_COMPLETED]: AutoAction
+  [LifeCycleEventTypeAsConst.VIEW_CREATED]: ViewCreatedEvent
+  [LifeCycleEventTypeAsConst.VIEW_UPDATED]: ViewEvent
+  [LifeCycleEventTypeAsConst.VIEW_ENDED]: ViewEndedEvent
+  [LifeCycleEventTypeAsConst.REQUEST_STARTED]: RequestStartEvent
+  [LifeCycleEventTypeAsConst.REQUEST_COMPLETED]: RequestCompleteEvent
+  [LifeCycleEventTypeAsConst.SESSION_EXPIRED]: void
+  [LifeCycleEventTypeAsConst.SESSION_RENEWED]: void
+  [LifeCycleEventTypeAsConst.PAGE_EXITED]: PageExitEvent
+  [LifeCycleEventTypeAsConst.RAW_RUM_EVENT_COLLECTED]: RawRumEventCollectedData
+  [LifeCycleEventTypeAsConst.RUM_EVENT_COLLECTED]: RumEvent & Context
+  [LifeCycleEventTypeAsConst.RAW_ERROR_COLLECTED]: {
     error: RawError
     savedCommonContext?: CommonContext
     customerContext?: Context

--- a/scripts/cli
+++ b/scripts/cli
@@ -78,6 +78,7 @@ cmd_check_typescript_compatibility () {
   cd test/app
   rm -rf node_modules
   yarn
+  check_typescript_isolated_modules_compatibility || fail 'typescript isolated modules compatibility broken'
   check_typescript_3_compatibility || fail 'typescript@3.8.2 compatibility broken'
   check_typescript_latest_compatibility || fail 'typescript@latest compatibility broken'
 }
@@ -88,7 +89,7 @@ check_typescript_3_compatibility () {
 
 check_typescript_latest_compatibility () {
   # Add typescript options only supported in newer versions
-  perl -i -l -p -e 'print "\"exactOptionalPropertyTypes\": true," if $. == 3' tsconfig.json
+  insert_line_in_file tsconfig.json 3 '"exactOptionalPropertyTypes": true,'
 
   yarn add --dev typescript@latest
   local succeeded=true
@@ -100,6 +101,27 @@ check_typescript_latest_compatibility () {
   git checkout .
 
   $succeeded
+}
+
+check_typescript_isolated_modules_compatibility () {
+  insert_line_in_file tsconfig.json 3 '"isolatedModules": true,'
+
+  local succeeded=true
+  if ! yarn compat:tsc; then
+    succeeded=false
+  fi
+
+  # Revert all changes
+  git checkout .
+
+  $succeeded
+}
+
+insert_line_in_file () {
+  local file="$1"
+  local line_number="$2"
+  local line_content="$3"
+  perl -i -l -p -e "print '${line_content}' if \$. == ${line_number}" "$file"
 }
 
 cmd_check_server_side_rendering_compatibility () {

